perl_siphash.h - add some documentation

demerphq opened the pull request:
This adds a seedable 64 bit variant for testing purposes (at Scott's request), as well as documentation which is suitable to generate output in perlintern.pod 

* This set of changes does not require a perldelta entry.

bram-perl commented on perl_siphash.h:
```diff
@@ -43,75 +119,82 @@ void S_perl_siphash_seed_state(const unsigned char * const seed_buf, unsigned ch
 PERL_STATIC_INLINE U64 \
 FNC ## _with_state_64 \
   (const unsigned char * const state, const unsigned char *in, const STRLEN inlen) \
-{                                           \
-  const int left = inlen & 7;               \
-  const U8 *end = in + inlen - left;        \
-                                            \
-  U64 b = ( ( U64 )(inlen) ) << 56;         \
-  U64 m;                                    \
-  U64 v0 = U8TO64_LE(state);                \
-  U64 v1 = U8TO64_LE(state+8);              \
-  U64 v2 = U8TO64_LE(state+16);             \
-  U64 v3 = U8TO64_LE(state+24);             \
-                                            \
-  for ( ; in != end; in += 8 )              \
-  {                                         \
-    m = U8TO64_LE( in );                    \
-    v3 ^= m;                                \
-                                            \
-    SIP_ROUNDS;                             \
-                                            \
-    v0 ^= m;                                \
-  }                                         \
-                                            \
-  switch( left )                            \
-  {                                         \
+{                                   \
+  const int left = inlen & 7;       \
```
Bunch of white space changes included, not sure if that was really intended to go in this commit

demerphq replied:
It was intended. But i have split out.

bram-perl commented on perl_siphash.h:
```diff
@@ -43,75 +119,82 @@ void S_perl_siphash_seed_state(const unsigned char * const seed_buf, unsigned ch
     return FNC ## _with_state((U8*)state,in,inlen);                 \
+}                           \
+                            \
+PERL_STATIC_INLINE U64      \
+FNC ## _64 (const unsigned char * const seed, const unsigned char *in, const STRLEN inlen) \
+{                           \
+    U64 state[4];           \
+    SIPHASH_SEED_STATE(seed,state[0],state[1],state[2],state[3]);   \
+    return FNC ## _with_state_64((U8*)state,in,inlen);              \
```
It looks like this function was added? (not 100% sure because the white space changes makes it a bit harder to see)

Again, not sure if this was intended or not.  (If it was then it might be good to mention it in the commit message since it's doing a bit more then "add some documentation")

demerphq replied:
Yes it was intended, but i have split it out from the doc change now.

demerphq committed:
perl_siphash.h - add a 64 bit variant for testing, cleanup line endings

Scott Baker wanted a 64 bit general purpose hash function. Currently our
Siphash implementation is optimized for use in the internal hash tables
by having its state initialization separated from the main hash
function. We have a 32 bit variant we use for testing that takes a seed
and not a state, but our only 64 bit support has the seeding step split
out. This patch exposes a seedable 64 bit function.

Note, that Siphash *is* a 64 bit hash function, however historically
we only use 32 bit hashes in our hash tables. The new variant we
expose here basically just exposes the normal Siphash that most people
would use.

A follow up patch will add documentation.

demerphq committed:
perl_siphash.h - add documentation

Scott Baker asked for documentation for these functions. This adds
documentation sections so that autodoc can include docs for these
functions. Includes usage examples.

scottchiefbaker commented:
Since part of the reason this came up was how to hash a U64 into a U64 can you add an example of how one would implement this?

demerphq commented:
> can you add an example of how one would implement this?

Done.

scottchiefbaker commented:
Thank you for adding examples and test vectors. I can't speak for the C, but the documentation looks excellent.
